### PR TITLE
Upgrade ES cluster

### PR DESCRIPTION
**Pull request by jtyoung**
#78

### CODE REVIEW

**Comment by joshfinnie** [comment]
@jtyoung where are we with this post? Any where close to being reviewed yet?

**Comment by joshfinnie** [comment]
@jtyoung am gonna post this. g2g?

**Comment by jtyoung** [comment]
@joshfinnie I see some grammar issues. Lemme edit it up in a bit and re-push.

File: content/upgrade-es-seamless.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+Our biggest problem was with disk space usage. Indexing our dataset took up far more space than we had anticpated, and we needed to increase the size of the attached AWS EBS volumes on our instances. Slightly less pressing than our disk usage was our heap usage, which made us want to increase the memory of each node so that we could allocate more heap space (reminder that it's [very important](https://www.elastic.co/guide/en/elasticsearch/guide/current/heap-sizing.html#compressed_oops) to keep your heap size below 30.5gb). All the while, we were several versions behind: running 1.3.2 at the time of 1.7.0's release. We figured the best course of action was to tackle all three of these problems at the same time, without any cluster downtime.
```
Comment (jtyoung) [inline content/upgrade-es-seamless.md]: anticipated*

Reword the "slightly less pressing" sentence.

File: content/upgrade-es-seamless.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+Some caveats before we begin:
+ - This only applies to minor version upgrades, like 1.x to 1.x, or 2.x to 2.x. Major version upgrades, like 1.x to 2.x, are more complicated and require a full cluster restart.
+ - If you only have one master-eligible node, then taking it down will make your cluster very sad. If you're using the more recommended number of 3 master-eligible nodes, a new one will be elected seamlessly as you take each down individually for upgrades
```
Comment (jtyoung) [inline content/upgrade-es-seamless.md]: You can do better than "more recommended"

File: content/upgrade-es-seamless.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+From a high level, our process looks like this:
+* Launch a new, larger EC2 instance based off our current node AMIs, with larger EBS volumes attached
+* Upgrade this instance's version of Elasticsearch
+* Create a new AMI for our upgraded instance
+* Join this new node to our production cluster
+* Re-allocate all shards off an older, non-upgraded node
+* Allow the cluster to rebalance itself
+* Once all nodes have been shipped off the old node, shut it down
+* Repeat this process of spawning new instances based off the upgraded AMI until all nodes have been upgraded
```
Comment (jtyoung) [inline content/upgrade-es-seamless.md]: Reword this. Perhaps number the steps and change this to "repeat steps X through Y until all nodes are upgraded"

File: content/upgrade-es-seamless.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+First, launch a new instance with the specs you want. It makes things easier if you base this instance off an image of a working Elasticsearch node, so that your config settings are retained and you do not forget to install any required dependencies. You don't want this node to join your cluster yet, as you still need to upgrade it. We prevented this by simply not giving it access to our Elasticsearch AWS security groups. Now, let's upgrade our version.
+
+Elasticsearch has good [documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/rolling-upgrades.html) around how to perform a rolling version upgrade. It's best to follow their steps closely. However, since the node we're upgrading aren't a part of our cluster yet, it's much simpler. We have Elasticsearch installed from their tar packages, so we download the latest version, and place it in a directory next to our current version. We also manage our Elasticsearch processes with `supervisor`, so we shut it down as well.
```
Comment (jtyoung) [inline content/upgrade-es-seamless.md]: since the node we're upgrading isn't*

File: content/upgrade-es-seamless.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+* Stop supervisor
+ - `sudo service supervisor stop`
+* Download and extract the new version
+ - `wget https://download.elastic.co/elasticsearch/elasticsearch/elasticsearch-1.7.0.tar.gz && tar xvzf elasticsearch-1.7.0.tar.gz && rm -f elasticsearch-1.7.0.tar.gz && sudo mv elasticsearch-1.7.0 /opt/elasticsearch-1.7`
+* Copy over your previous config settings
+ - `sudo rm /opt/elasticsearch-1.7/config/elasticsearch.yml`
+ - `sudo mv /opt/elasticsearch/config/elasticsearch.yml /opt/elasticsearch-1.7/config/`
```
Comment (jtyoung) [inline content/upgrade-es-seamless.md]: Perhaps explain these commands a bit better.

File: content/upgrade-es-seamless.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+* Delete the previous version and put the new version in its place
+ - `sudo rm -rf /opt/elasticsearch`
+ - `sudo mv /opt/elasticsearch-1.7 /opt/elasticsearch`
+* Restart supervisor
+ - `sudo service supervisor start`
+
+With this newly upgraded node, we create an AMI of this instance so that we don't need to keep performing this manual upgrade process. All new instances going forward should be based off this upgraded image.
+
+Now, it's time to join this node to the cluster. We add the necessary security groups and watch our cluster health (`localhost:9200/_cluster/health?pretty`) show us what's happening. We should see a new data node join the cluster, and our cluster health change from `green` to `yellow`. The `yellow` state happens because of our mismatched version numbers in the cluster. Primary shards assigned to the newer version will not allocate their replica shards to older versioned nodes. With only one upgraded node, we will have unassigned replica shards. This is remedied once we have two upgraded nodes in the cluster, and health will again return to `green`.
```
Comment (jtyoung) [inline content/upgrade-es-seamless.md]: Add a link to the ES docs stating why cluster health goes to yellow.

File: content/upgrade-es-seamless.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+* Re-allocate all shards off the node
+```
+curl -XPUT localhost:9200/_cluster/settings -d '{
+    "transient" :{
+        "cluster.routing.allocation.exclude._ip" : "<IP Address>"
+    }
+}'
+```
+* Wait for the cluster to re-balance itself by waiting for `relocating_shards` to go to 0
+ - `curl localhost:9200/_cluster/health?pretty`
```
Comment (jtyoung) [inline content/upgrade-es-seamless.md]: Reword to make clearer.

**Comment by jtyoung** [comment]
@joshfinnie Ready for one more look over!

**Pull request merged**